### PR TITLE
Arbitrary clipping planes for volumes in the image layer #2

### LINKED ISSUE #3079: ensure_sequence_of_iterables : ValueError for empty list.
## 🐛 Bug

While chasing down some problems with displaying multi-channel, pyramidal ome.zarr files through a plugin
I noticed a problem in `ensure_sequence_of_iterables`:

https://github.com/napari/napari/blob/3bd58bde6858eb4d9f916fe32df262a4ebe1a24f/napari/utils/misc.py#L151

When the `obj` passed in is an empty list `[]`, an index error is raised:

```
~/Github/napari/napari/utils/misc.py in ensure_sequence_of_iterables(obj=[], length=4)
    182     """
    183     print(f"in ensure sequence of iterables {obj}")
--> 184     if obj is not None and is_sequence(obj) and is_iterable(obj[0]):
        obj = []
        global is_sequence = <function is_sequence at 0x7f45f5aa8550>
        global is_iterable = <function is_iterable at 0x7f45f5aa84c0>
    185         if length is not None and len(obj) != length:
    186             raise ValueError(

IndexError: list index out of range
```

In the ideal world (no upstream errors) there probably shouldn't be empty lists passed into this function. 
However, it is not quite clear to me what the desired failure mode is for (possibly incorrect) input such as
`[]` or `None`. For `[]` one gets the above `IndexError`, for `None` one will get an iterator that will repeatedly 
output `None`.
I am not sure whether the latter (repeated `None` ) is desired behaviour or an error should be raised.

### PR DESCRIPTION
**Pull request by brisvag**
# Description
This PR is a remake of #2934. It uses the [updated vispy PR (now merged!)](https://github.com/vispy/vispy/pull/2116).

I shamelessly copied some code from @alisterburt 's example for plane rendering, so now I have a beatuful interactive example too (edit: now with labels!) :)

https://user-images.githubusercontent.com/23482191/129874843-b14aa93c-da1d-40b9-9044-b6dabc2083ff.mp4

Here, I tried to implement something similar to #3023, using a `PlaneManager` to do the work. I'm not sure how higher level interface should be, though... Suggestions are welcome!

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# References
- #2934
- #3023
- [vispy#2116](https://github.com/vispy/vispy/pull/2116)

# How has this been tested?
- [x] added/updated tests where necessary

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/_vispy/vendored/volume.py
Comment (alisterburt) [inline napari/_vispy/vendored/volume.py]:
```suggestion
// Only sample volume if loc is not clipped by clipping planes
```

File: napari/_vispy/vendored/volume.py
Comment (alisterburt) [inline napari/_vispy/vendored/volume.py:980]: out of interest, do you have a sense for how performant this is? Does it scale well with the number of clipping planes?

File: napari/layers/utils/plane_manager.py
Comment (alisterburt) [inline napari/layers/utils/plane_manager.py]: might be worth making this a `SelectableEventedList`, to simplify dealing with 'which plane am I dealing with', what do you think?

Comment (alisterburt) [inline napari/layers/utils/plane_manager.py]: it might be nice to create a convenience function/method for creating a `PlaneList` for an axis aligned bounding box (6 planes) centered around a position in a volume with certain dimensions in 0,1,2, obviously this isn't necessary for this PR

File: napari/layers/utils/plane_manager.py
Comment (alisterburt) [inline napari/layers/utils/plane_manager.py]: thicc

File: napari/layers/utils/plane_manager.py
Comment (alisterburt) [inline napari/layers/utils/plane_manager.py]: this is nice!

**Review by alisterburt — COMMENTED** [review]
Awesome!! Really excited to get this in!

Left a few comments, one thing that feels 'missing' is that the `PlaneManager` objects have an `enabled` attribute which doesn't affect whether or not those clipping planes are active once instantiated, only on instantiation. You could connect the `PlaneManager.events.enabled` events to the `VispyImageLayer.on_clipping_planes_change()` method to update the array which gets passed to the shader?

Otherwise my main suggestion is to make the `PlaneList` a `SelectableEventedList` to make moving through and highlighting clipping planes a bit simpler when we build interactivity onto this a bit more

So cool!!

File: napari/_vispy/vendored/volume.py
Comment (brisvag) [reply napari/_vispy/vendored/volume.py:980]: Do you mean how performant caching is, or how performant many clipping planes are? I expect caching to be negligible, and the number of clipping planes to be more important. However, as mentioned upstream, it's probably fair to assume clipping planes won't be used in amounts bigger than ~10, at which point you can probably merge 2 almost-equal planes into 1 and not even notice.

Comment (alisterburt) [reply napari/_vispy/vendored/volume.py:980]: I meant the performance of the clipping planes in general! I can actually imagine us using this relatively heavily - given that we may use 6 planes for a bounding box and may want to add more... can cross that bridge when we come to it though :)

File: napari/layers/utils/plane_manager.py
Comment (brisvag) [reply napari/layers/utils/plane_manager.py]: Totally! I was planning to do that already in a follow up PR, but first I want to be sure on how these models should look like here.

File: napari/_vispy/vendored/volume.py
Comment (brisvag) [reply napari/_vispy/vendored/volume.py:980]: 6 for a box + maybe 4 for additional clipping is 10 :P 2 boxes seems already redundant... But yeah, I don't expect this to be problematic any time soon!

File: napari/layers/utils/plane_manager.py
Comment (brisvag) [reply napari/layers/utils/plane_manager.py]: That's a good idea, would make interactivity easier!

File: napari/layers/utils/plane_manager.py
Comment (brisvag) [reply napari/layers/utils/plane_manager.py]: To answer you top comment: `plane.enabled` works as you might expect! Since I connected the whole `EventedList` to the update method, this will fire for any `PlaneManager` event, including changing `plane.enabled`. Try to run the example from ipython, then:
```python
volume_layer.clipping_planes[0].enabled = False
```

Comment (alisterburt) [reply napari/layers/utils/plane_manager.py]: doh! Should've checked - that's awesome :)

**Comment by alisterburt** [comment]
weird visual bug, investigating 
image

File: napari/_vispy/vendored/volume.py
Comment (alisterburt) [inline napari/_vispy/vendored/volume.py]:
```suggestion
            // Only sample volume if loc is not clipped by clipping planes
            float is_shown = $clip_by_planes(loc, u_shape);
```

**Comment by alisterburt** [comment]
fixed, PRing now
image

File: napari/layers/utils/plane_manager.py
Comment (brisvag) [reply napari/layers/utils/plane_manager.py]: I decided to do this, since it wasn't too hard. Try this on the example:
```python
volume_layer.clipping_planes = volume_layer.clipping_planes.from_bounding_box([32, 32, 32], [40, 20, 30]) # (center, dimensions)
```

**Review by alisterburt — COMMENTED** [review]
Tests are failing because of the way you are trying to serialise the clipping planes

https://github.com/napari/napari/pull/3140/files#diff-9e4cd2702cfbb10e85a517ed53426c08d163353b58aab0e13609b4375c1e9780R848

To keep the events working as expected with serialisation we probably want to keep the same `EventedList` and just repopulate it `PlaneList.clear()`, `PlaneList += new_plane_list` when setting with a `PlaneList`

**Comment by brisvag** [comment]
I was just working on it, modeling on your #3143 :)

File: napari/utils/misc.py
Comment (brisvag) [inline napari/utils/misc.py]: I added this line because serialization was failing for `clipping_planes`. Apparently, until now the case of "empty list as argument" wasn't considered. Hopefully I didn't break anything with this (tests pass, but I'd like confirmation from others that this makes sense).

**Comment by brisvag** [comment]
[This test run is very strange](https://github.com/napari/napari/actions/runs/1120463668). As far as I can tell, there is maybe an issue with how the docstring is parsed (resulting "sometimes" in a parameter name that includes what comes after the `:` O.o). Somehow, this does not happen in python 3.9 (which is my local env). I'd love some suggestions :/

**Comment by sofroniewn** [comment]
Just popping by to say this looks very exciting!! Let me know when ready for review and i'll check it out

**Comment by brisvag** [comment]
@sofroniewn I think this is pretty much ready, except for this weird test error. I can't figure out how to debug it!

File: napari/layers/utils/plane_manager.py
Comment (sofroniewn) [inline napari/layers/utils/plane_manager.py]: I wonder if we need a new object here, or if we should just allow `thickness = None` for clipping planes. I worry that the additional object causes a bit more overhead when reasoning about things for not too much gained. Thoughts @jni?

File: napari/utils/misc.py
Comment (sofroniewn) [reply napari/utils/misc.py]: @tlambert03 any further thoughts here?

**Review by sofroniewn — COMMENTED** [review]
Ok this is coming along very nicely - my main feedback is I don't know if we need to create the `ThickPlaneManager` object, I think we can just have `thickness=None` for clipping planes.

This PR also needs to be rebased, since the last ones merged.

I think it might nice be nice to get #3150 in first too and but either order is also fine.

Great work here. I'll review once more when ready

File: napari/utils/misc.py
Comment (tlambert03) [reply napari/utils/misc.py]: actually, this is https://github.com/napari/napari/issues/3079  (So you could mark this as closing that issue when you fix it).  copied from my comment there:

> I think the behavior for an empty list should be to return an empty list (unless length is provided and not 0, in which case it should be a valueError

so rather than the current implementation (which will return `repeat([])` when provided an empty list), let's just return [] .... _unless_ length is not None, in which case it should assert the length matches

File: napari/layers/utils/plane_manager.py
Comment (tlambert03) [reply napari/layers/utils/plane_manager.py]: I also wonder about the naming "manager" ... i see we're calling more and more things "Manager".  What is this managing?  it looks like a pretty simple set of attributes describing a plane, in which case I'd propose we call it `Plane` ... and I agree with @sofroniewn that it doesn't need a new subclass.  I'd vote for `Plane` having a default thickness of 0, and specifying thickness > 0 when you need a "ThickPlane".

Comment (sofroniewn) [reply napari/layers/utils/plane_manager.py]: Yup agreed - these are just planes, let's rename `Plane` and go with default thickness of `0`. Thanks for input @tlambert03!!

Comment (alisterburt) [reply napari/layers/utils/plane_manager.py]: I switched to manager when I added the `enabled` attribute because it's an object managing the plane rendering but I agree it's not really necessary, let's keep it simple!

Comment (tlambert03) [reply napari/layers/utils/plane_manager.py]: Yeah, I think we should reserve "manager" for something managing a set of objects (like the old "settings manager" was managing a set of more complicated objects).  If  it's just "managing" a set of attributes that are used somewhere, then i think it's pretty much a pure model object.

Comment (brisvag) [reply napari/layers/utils/plane_manager.py]: Agree on everything!

**Comment by brisvag** [comment]
While I fix these few things, I'd like to direct everyone who missed it to #3148. Maybe it's better to discuss there before merging this PR.

File: napari/utils/misc.py
Comment (brisvag) [reply napari/utils/misc.py]: I think we have a problem here: as far as I can tell, `repeat([])` *is* the right value in the clipping plane case (`[]` is the argument value, and it should be repeated for each channel). However, as you said in #3079, sometimes `[]` may mean an *empty* sequence of iterables. How do we solve this ambiguity?

Comment (tlambert03) [reply napari/utils/misc.py]: I feel like that should be up to the consumer of this function.  `[]` is a valid sequence of iterables (of len 0), can you show me where the clipping plane is using this function?

Comment (brisvag) [reply napari/utils/misc.py]: If you pass `clipping_planes=[]` (a valid agument) to `add_image` in multi-channel mode (let's say with 3 channels), `ensure_sequence_of_iterables` will be called to determine if and how this can be done. If the function returns simply `[]` instead of `repeat([])`, all sorts of `IndexError`s pop up, trying to acces the 3 expected elements of the return value.

Which are the cases where `[]` interpreted as an empty sequence of iterables is desirable? (0 channels?)

Comment (tlambert03) [reply napari/utils/misc.py]: just curious, why does `clipping_planes=[]` have to be a valid argument in multi-channel mode?  Kinda seems like `None` is how we're doing "null values" with all the other things.

Comment (tlambert03) [reply napari/utils/misc.py]: i don't have a terribly strong opinion here... so if you want to just make it `repeat([])` and move on with your life, that's fine.

when using multi-channel mode, `ensure_sequence_of_iterables` will be called with a `length` parameter, and `[]` would _not_ pass that requirement there, so i do think it feels a little bit odd for a valid null value with a specified length.  but since that function is mostly only used for `add_image` anyway, i guess we can make it be what that consumer needs?

Comment (brisvag) [reply napari/utils/misc.py]: I think `[]` should be a valid argument because it is valid in single-channel mode, and having inconsistent behaviour would be different from all the other layer parameters (I think?). `None` *is* valid as well, and poses no problem. I guess I could change it to use/accept only `None` when the `PlaneList` is empty, and refuse `[]`? That seems a bit counterintuitive to me, but it would solve this issue.

> when using multi-channel mode, ensure_sequence_of_iterables will be called with a length parameter, and [] would not pass that requirement there

Yeah, I noticed that as well. My current solution (WIP) is to add `repeat_empty` as a parameter to the function, default to false. This way, the behaviour stays as in main and the function can be usel elsewhere as previously indended. However, in the multichannel call, we can set `repeat_empty` to true and get this different behaviour. This plays well with `length`:
```python
In [37]: ensure_sequence_of_iterables([])
Out[37]: []

In [38]: ensure_sequence_of_iterables([], length=1)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-38-e8e5de8d471d> in <module>
----> 1 ensure_sequence_of_iterables([], length=1)

~/git/napari/napari/utils/misc.py in ensure_sequence_of_iterables(obj, length, repeat_empty)
    194             if (len(obj) == 0 and not repeat_empty) or len(obj) > 0:
    195                 # sequence of iterables of wrong length
--> 196                 raise ValueError(
    197                     trans._(
    198                         "length of {obj} must equal {length}",

ValueError: length of [] must equal 1

In [39]: ensure_sequence_of_iterables([], repeat_empty=True)
Out[39]: repeat([])

In [40]: ensure_sequence_of_iterables([], length=1, repeat_empty=True)
Out[40]: repeat([])
```

What do you think?

Comment (brisvag) [reply napari/utils/misc.py]: I also have doubts about this:

https://github.com/napari/napari/blob/aa815ea0430232cb91f7d8111da68c5a880841d3/napari/utils/_tests/test_misc.py#L31-L32

The test seems to imply that this is intended behaviour. Why is is so? This way `((0, 1), None, None)` is treated differently from `(None, (0, 1), None)`, which seems wrong to me. As far as I can tell, both of these tests should result in `repeat((...))`, as they are not sequences of iterables.

I currently changed the function to use `if all(is_iterable(el) for el in obj)` rather than `if len(obj) > 0 and is_iterable(obj[0])`, which results in what I expected, but maybe I'm missing something.

File: napari/_vispy/vendored/volume.py
Comment (brisvag) [inline napari/_vispy/vendored/volume.py:394]: @kevinyamauchi for reference, these are the lines we are talking about over at #3181 regarding the `discard` call.

**Comment by brisvag** [comment]
Several new changes. Notably:
- `PlaneManager` is now a single object (no more `ThickPlaneManageer`)
- `slicing_plane` and `clipping_planes` now are `experimental_` (I had to add/fix some things compared to #3150)
- changed a few things to fix #3079, as mentioned in the comments above.

File: napari/_vispy/vendored/volume.py
Comment (alisterburt) [inline napari/_vispy/vendored/volume.py:12]: are these extra spaces there in the vispy shader? good to try to keep things in sync there - no major issue though

File: examples/clipping_planes_interactive.py
Comment (alisterburt) [inline examples/clipping_planes_interactive.py:138]: for these interactive examples I've moved to adding a small explanation in the [`Viewer.text_overlay`](https://github.com/napari/napari/blob/master/examples/viewer_fps_label.py) - could do the same here but not strictly necessary!

File: napari/_vispy/vendored/volume.py
Comment (alisterburt) [inline napari/_vispy/vendored/volume.py:980]: What's the expected/actual behaviour here if more than 10 clipping planes are added? If it doesn't work, this info should be added to a docstring somewhere

File: napari/_vispy/vendored/volume.py
Comment (alisterburt) [inline napari/_vispy/vendored/volume.py:1009]: should mention which side of the normal is kept and which is discarded, also should be clear about which index on axis 1 is for plane points and which for plane normals

**Review by alisterburt — APPROVED** [review]
This looks good to me, left a few small comments but I would happy to merge this as is! Awesome job @brisvag 🎉

File: napari/layers/utils/plane.py
Comment (sofroniewn) [inline napari/layers/utils/plane.py]: Should we make default thickness be 0? 10 seems a little random.

File: napari/layers/utils/plane.py
Comment (sofroniewn) [inline napari/layers/utils/plane.py]: Maybe say something about a thickness of zero being used for slicing planes

File: napari/layers/utils/plane.py
Comment (sofroniewn) [inline napari/layers/utils/plane.py]: Should we enable by default? I feel like if you're making it, you probably want it enabled

File: napari/layers/utils/plane.py
Comment (sofroniewn) [inline napari/layers/utils/plane.py]: Strictly these are probably better thought of as sliced data coordinates as they are 3D and not nD.

File: napari/layers/utils/plane.py
Comment (sofroniewn) [inline napari/layers/utils/plane.py]: Strictly these are probably better thought of as sliced data coordinates as they are 3D and not nD.

**Review by sofroniewn — APPROVED** [review]
Looks great! Left some mostly minor comments. Example works great for me. Excited to get this in and keep this work moving forward!!

File: napari/_vispy/vendored/volume.py
Comment (brisvag) [reply napari/_vispy/vendored/volume.py:12]: The spaces were removed by my vispy PR already, so now that it's merged everything should be in sync!

**Comment by alisterburt** [comment]
Sweet!

Edit: this was a reply to a specific comment

File: examples/clipping_planes_interactive.py
Comment (brisvag) [reply examples/clipping_planes_interactive.py:138]: I have tried to set a value and set it to visible, but nothing appears. Am I missing something?
```python
viewer.text_overlay.text = 'test'
viewer.text_overlay.visible = True
```

File: napari/_vispy/vendored/volume.py
Comment (brisvag) [reply napari/_vispy/vendored/volume.py:980]: It works just fine. This cache holds up to 10 code snippets, each one generated to allow a different amount of clipping planes. It does not limit the amount of clipping planes! I just means that if you (in the same session) clip with 1 plane, then 2, then 10, then 22, and so on up to 10 different numbers, the 11th time it will discard the first cached code snippet.

The reasoning behind it is that in normal use people will tend to increase/decrease clips by one at a time and probably stay under 10 clips in total, so caching the first ten should cover the typical use case of switching back and forth between (say) 3-4-5 planes.

File: napari/_vispy/vendored/volume.py
Comment (brisvag) [reply napari/_vispy/vendored/volume.py:1009]: Uhm, this should probably be added upstream as well... Will do here on the napari side for now!

File: napari/layers/utils/plane.py
Comment (brisvag) [reply napari/layers/utils/plane.py]: Oops, copy-paste mistake!

File: napari/layers/utils/plane.py
Comment (brisvag) [reply napari/layers/utils/plane.py]: With clipping planes, true, and I agree in general this class should instantiate enabled planes. However, with @alisterburt 's `slicing_plane` we want it disabled by default (since it exists on the layer from instantiation). Easy to solve here (just add enabled=True there), but this might change depending on how we move forward with #3148 so it's worth noting.

File: napari/layers/utils/plane.py
Comment (brisvag) [reply napari/layers/utils/plane.py]: You probably mean clipping planes? I would not put it here, because
1) the plane is agnostic about what it's used for
2) thickness is actually unused by clipping planes, so it could be set to anything.
3) We discussed that we could maybe make clipping planes behave as a soft-edge mask when thickness is above 0.

If anything, I'd rather mention this in the `Image` and `Label` layer docstrings about clipping planes, saying that thickness is unused. What do you think?

File: napari/layers/utils/plane.py
Comment (alisterburt) [reply napari/layers/utils/plane.py]: I think there's definitely an argument for having a plane class which we subclass as necessary for the different types (`SlicingPlane`, `ClippingPlane`) - thoughts? I know subclassing can be messy but this seems like the best way to go to me

File: napari/layers/utils/plane.py
Comment (alisterburt) [reply napari/layers/utils/plane.py]: see point below about subclassing specific types of planes to simplify/remove the need for these sorts of discussions

File: examples/clipping_planes_interactive.py
Comment (alisterburt) [reply examples/clipping_planes_interactive.py:138]: hmm, I always do this in the other order (visible then change the text), might be a bug?

**Comment by brisvag** [comment]
I added a labels layer to the example,  cause I realized I hadn't tested layer clipping so far 😅, luckily it works!.

And it seems that @kevinyamauchi 's 3D rendering of labels is playing well with clipping planes!

https://user-images.githubusercontent.com/23482191/129874843-b14aa93c-da1d-40b9-9044-b6dabc2083ff.mp4

**Comment by alisterburt** [comment]
that's awesome @brisvag - looks so good!! Can you check what happens to the grey edges on labels if you have a clipping plane just a bit inside the volume? I'm wondering if we can use this to get rid of the grey edge effects
edit: saw in the example that it won't, never mind!

File: napari/layers/utils/plane.py
Comment (brisvag) [reply napari/layers/utils/plane.py]: That's what I did before, and [people seemed to agree that it overcomplicates things](https://github.com/napari/napari/pull/3140#discussion_r688972201). Happy to go either way, but now I kinda like the fact that `Plane` is agnostic regarding its application.

File: examples/clipping_planes_interactive.py
Comment (brisvag) [reply examples/clipping_planes_interactive.py:138]: Yup, seems like it!

Comment (brisvag) [reply examples/clipping_planes_interactive.py:138]: Actually, I can't reproduce... I must have messed up something?

**Comment by alisterburt** [comment]
Looks like this is ready to go! I'll merge tomorrow morning unless anybody raises concerns

File: napari/layers/utils/plane.py
Comment (tlambert03) [reply napari/layers/utils/plane.py]: > subclass as necessary for the different types 

Yeah, I guess the question would be what's different for those types.  If it's just methods (not attributes), then like @brisvag says, there's an argument for keeping the model free of the business logic... and letting the thing that has the plane use it to clip, or whatever.  

But, I guess it depends on how they're different

Comment (alisterburt) [reply napari/layers/utils/plane.py]: You're right that it's the same, removing the `Manager` and using the names `ClippingPlane`/`SlicingPlane` makes it feel a bit less 'imposing' for some reason - I'm happy whichever way this goes but do feel like these objects are different enough to warrant their own objects. 

I would be a little confused if I found a `Plane` object in the `ClippingPlaneList` which had a thickness and nothing happened when I changed it.

Another argument for separating them out is if we keep the reusable business logic in utility functions (which I support) we would have to differentiate between functions which operate on different *types* of the same object

For example
`intersect_line_with_plane()` (no thickness) and `intersect_line_with_slicing_plane` (with thickness), choosing the correct function as a user/dev requires knowledge about the other type of `Plane`, feels less obvious?

Comment (brisvag) [reply napari/layers/utils/plane.py]: I think we should keep in mind future plans:
1) ideas in #3148 do not play well with different objects
2) thickness will be useful for clipping as well if/when we implement soft clipping.

Comment (tlambert03) [reply napari/layers/utils/plane.py]: > do feel like these objects are different enough to warrant their own objects.

well you've been using these objects much more than me, so I trust your intuition here.

I think this touches on a broader interesting topic about architecture.  I _think_ (though I'm not sure) that it would be nice if we had "largely declarative" models that were acted upon by rendering engines.  Layers are in many cases just a set of attributes, and it's up to the vispy classes (view layers) to turn those attributes into meaning. 

> we would have to differentiate between functions which operate on different types of the same object

This is a good argument.  I think there's a couple options here.  One is that you don't have both `intersect_line_with_plane()` (no thickness) and `intersect_line_with_slicing_plane` (with thickness) ... you just have `intersect_line_with_plane` and it checks if the plane has thickness.  

But, if you have a whole family of functions that _require_ thickness (and have no meaning without it), or a whole set of functions that expect the output of `intersect_line` to be another line, rather than a point... then I definitely agree with you that those are fundamentally different models that require different classes.   
(Indeed, a true plane and a "slab of space" are fundamentally different types).  And I'd soften/retract my comment from https://github.com/napari/napari/pull/3140#discussion_r688973818 

@sofroniewn?

Comment (sofroniewn) [reply napari/layers/utils/plane.py]: I'd be fine with `ClippingPlane` and `SlicingPlane` as two class names, and then `ClippingPlaneList`. Feels pretty natural and intuitive in a way that `PlaneManager` and `ThickPlaneManager` didn't before as @alisterburt noted above. 

If you'd like to move that over tomorrow before merge that's fine. @alisterburt and @brisvag you can decided this between yourselves during your timezone and feel free to have this PR merged by the time I get back online up in ~10 hours :-)

Comment (alisterburt) [reply napari/layers/utils/plane.py]: Thanks for the input @tlambert03 @sofroniewn! I definitely like the idea of declarative models in general but dislike the 'forced' code separation that results: having functions which act on models rather than methods on objects forces users to think 'oh, where do I import that thing from again?'

I think this is my main argument against fully declarative models in a system that is designed to be easy for people to interact with programmatically... all of that to say 'I'm not sure which way I prefer overall' 🙂 

There are definitely a family of functions which require thickness, any 3d spatial stuff`point_in_slab_of_space()`/`volume_of_slab_of_space()` jumps to mind but I'm sure there are others

I'm going to PR this with the `ClippingPlane`/`SlicingPlane` changes soon, will merge once @brisvag has had a look at that

Comment (tlambert03) [reply napari/layers/utils/plane.py]: >  'oh, where do I import that thing from again?'
>  I think this is my main argument against fully declarative models in a system that is designed to be easy for people to interact with programmatically..

oh definitely... one shouldn't have to import every utility function that they need.
but I don't think the two concepts are necessarily that mutually exclusive.  It's true that someone couldn't necessary use a method on a deeply nested object:  `layer.something.something.some_method()`, but that's where our main API design should come in and let someone use `layer.something()` ... which then dispatches to the appropriate utility functions (so that people don't have to remember where to import them from)

_or_ we start thinking outside pattern of having layers do everything, and start introducing new objects that manage these relationships between layers (i.e., a couple clipping planes and a couple layers)... and _that_ manager has an API that does all this.  (instead of each layer remembering the clipping planes that it relates to, and directly calling the methods on those plane objects).

>I'm going to PR this with the ClippingPlane/SlicingPlane changes soon, will merge once @brisvag has had a look at that

sounds good!  onward :)

**Comment by sofroniewn** [comment]
> I'm going to PR this with the ClippingPlane/SlicingPlane changes soon, will merge once @brisvag has had a look at that

I believe we are now waiting for this correct @alisterburt - let me know when it's done and i can do a final pass on this PR and merge, thanks!!

File: napari/layers/utils/plane.py
Comment (brisvag) [reply napari/layers/utils/plane.py]: I still don't see the benefit of this split (`points_ins_slab_of_space` works just fine with thickness=0...), especially when considering the possible plans of using the same plane to (for example) clip one layer and slice another; with 2 different obects, we are forced to have both of them, and provide an api (and UI!) to "connect" them so they update together.

Anyways, I don't think this is important enough at the moment to hold up the merge, so we can go with the split for now. Just writing down my thoughts for when we revisit this!

**Comment by brisvag** [comment]
Merged @alisterburt 's changes, with some tweaks on code duplication and several docstring things cattered around the codebase (this image/labels/add_image docstring duplication is a bit of a pain :P)

**Comment by sofroniewn** [comment]
Ok thanks! Let's go with this now, we can refactor as we do more with clipping. Great to get this in now, thanks for all the hard work here folks!!

**Pull request merged**
